File: Hito_4/Observer/VehicleRoutingOriginal/VehicleRoutingSolucionadoObserver/VehicleRoutingValidator/Handlers/NoLocationIsVisitedTwiceHandler.cs
using VehicleRoutingValidator.VehicleRouting;

namespace VehicleRoutingValidator.Handlers;

public class NoLocationIsVisitedTwiceHandler : BaseHandler
{
    private const string Description = "Some locations are visited multiple times.";
    private const int Depot = VehicleRoutingProblem.Depot;
    private int[] _solution;

    public override bool IsSolutionValid(int[] solution)
    {
        _solution = solution;
        if(IsEveryLocationVisitedAtMostOneTime())
            return base.IsSolutionValid(_solution);
        UpdateSubscribers(Description);
        return false;
    }

    private bool IsEveryLocationVisitedAtMostOneTime()
    {
        for (int i = 0; i < _solution.Length; i++)
            if(IsNotDepot(i) && IsThisLocationVisitedTwice(i))
                return false;
        return true;
    }

    private bool IsThisLocationVisitedTwice(int idLocation)
    {
        for (int j = idLocation + 1; j < _solution.Length; j++)
            if (_solution[idLocation] == _solution[j])
                return true;
        return false;
    }

    private bool IsNotDepot(int idLocation)
        => _solution[idLocation] != Depot;
}

File: Hito_4/Observer/VehicleRouting/handlers/NoLocationIsVisitedTwiceHandler.py
import BaseHandler
from ..models.VehicleRoutingProblem import VehicleRoutingProblem

class NoLocationIsVisitedTwiceHandler(BaseHandler):
    DESCRIPTION = "Some locations are visited multiple times."
    DEPOT = VehicleRoutingProblem.DEPOT

    def __init__(self):
        super().__init__()
        self._solution = []

    def is_solution_valid(self, solution: list[int]) -> bool:
        self._solution = solution
        if self._is_every_location_visited_at_most_once():
            return super().is_solution_valid(solution)
        self.update_subscribers(self.DESCRIPTION)
        return False

    def _is_every_location_visited_at_most_once(self) -> bool:
        for i in range(len(self._solution)):
            if self._is_not_depot(i) and self._is_location_visited_twice(i):
                return False
        return True

    def _is_location_visited_twice(self, id_location: int) -> bool:
        for j in range(id_location + 1, len(self._solution)):
            if self._solution[id_location] == self._solution[j]:
                return True
        return False

    def _is_not_depot(self, id_location: int) -> bool:
        return self._solution[id_location] != self.DEPOT
